data_steward/vocabulary.py

```diff
@@ -1,11 +1,179 @@
-CONCEPT = 'concept'
-CONCEPT_ANCESTOR = 'concept_ancestor'
-CONCEPT_CLASS = 'concept_class'
-CONCEPT_RELATIONSHIP = 'concept_relationship'
-CONCEPT_SYNONYM = 'concept_synonym'
-DOMAIN = 'domain'
-DRUG_STRENGTH = 'drug_strength'
-RELATIONSHIP = 'relationship'
-VOCABULARY = 'vocabulary'
-VOCABULARY_TABLES = [CONCEPT, CONCEPT_ANCESTOR, CONCEPT_CLASS, CONCEPT_RELATIONSHIP, CONCEPT_SYNONYM, DOMAIN,
-                     DRUG_STRENGTH, RELATIONSHIP, VOCABULARY]
+import os
+import csv
+import re
+import sys
+import warnings
+
+from resources import CONCEPT, VOCABULARY, AOU_GENERAL_PATH, AOU_GENERAL_CONCEPT_CSV_PATH, hash_dir
+
+csv.field_size_limit(sys.maxsize)
+
+DELIMITER = '\t'
+LINE_TERMINATOR = '\n'
+RAW_DATE_REGEX = r'\d{8}$'  # yyyymmdd
+BQ_DATE_REGEX = r'\d{4}-\d{2}-\d{2}$'  # yyyy-mm-dd
+
+TRANSFORM_FILES = 'transform_files'
+APPEND_VOCABULARY = 'append_vocabulary'
+APPEND_CONCEPTS = 'append_concepts'
+ADD_AOU_GENERAL = 'add_aou_general'
+ERRORS = 'errors'
+
+AOU_GEN = 'AoU_General'
+AOU_GEN_VOCABULARY_CONCEPT_ID = '2000000000'
+AOU_GEN_VOCABULARY_REFERENCE = 'https://docs.google.com/document/d/10Gji9VW5-RTysM-yAbRa77rXqVfDfO2li2U4LxUQH9g'
+OMOP_VOCABULARY_CONCEPT_ID = '44819096'
+
+ERROR_APPENDING = 'Appending to {in_path} which already contains rows for ' + AOU_GEN
+
+
+def format_date_str(s):
```

Comment:
I would change this function to use the parse function from dateutil.parser .  I would note in the documentation string that I expect this to raise a ValueError.

```
:raises:  ValueError if a valid date object cannot be parsed from the string
"""
date_obj = parse(date_str)
formatted_date_str = date_obj.strftime('%Y-%m-%d')
return formatted_date_str
```

This will let you get rid of those DATE regular expressions.

Reply:
This is great advice! Done.

Reply:
I wanted to note that the performance can take a hit here:
`timeit.timeit('format_date_str_regex(s)', setup="from __main__ import format_date_str_regex; s='20170517'", number=5000)`
0.0126650333405
`timeit.timeit('format_date_str_strftime(s)', setup="from __main__ import format_date_str_strftime; s='20170517'", number=5000)`
0.318553924561

I think we should generally stick to standard libraries for dates but in this case regex is much faster and we could make an exception. I ran into this issue while trying to load the new vocabulary. The millions of entries make a big difference.

Reply:
Good catch Nishanth.  Thanks!  I had not run a time comparison on this.  With you suggestion I ran the following:
```
timeit.timeit("from dateutil.parser import parse\ndate_obj = parse(date_str)\nformatted = date_obj.strftime('%Y-%m-%d')", setup = 'date_str = "20170101"', number = 1000000)
timeit.timeit("if bq_pattern.match(date_str):\n    formatted = date_str\nelif raw_pattern.match(date_str):\n    parts = date_str[0:4], date_str[4:6], date_str[6:8]\n    formatted = '-'.join(parts)", setup="import re\ndate_str='20170101'\nRAW_DATE_REGEX=r'\d{8}$'\nBQ_DATE_REGEX = r'\d{4}-\d{2}-\d{2}$'\nraw_pattern = re.compile(RAW_DATE_REGEX)\nbq_pattern=re.compile(BQ_DATE_REGEX)", number = 1000000)
timeit.timeit("if re.match(BQ_DATE_REGEX, date_str):\n    formatted = date_str\nelif re.match(RAW_DATE_REGEX, date_str):\n    parts = date_str[0:4], date_str[4:6], date_str[6:8]\n    formatted = '-'.join(parts)", setup="import re\ndate_str='20170101'\nRAW_DATE_REGEX=r'\d{8}$'\nBQ_DATE_REGEX = r'\d{4}-\d{2}-\d{2}$'\n", number = 1000000)
```
I rant this with a pre-compiled regular expression and one that is compiled every time.  The string format one in 52.6 seconds, the regular expression one ran in 1.7 seconds, and the pre-compiled regular expression one in 0.7 seconds.  For performance, regular expressions are better.  It's a personal preference, but I prefer to pre-compile my regular expressions when I use them..

Reply:
I've been thinking about this.  Yes, the `dateutil.parser` is much slower than regular expressions.  It was the `parse` function that raised the exception with the bad date objects in the vocabulary files (00170101 dates).  If we use a regular expression instead of the `parse` function, we should consider making the regular expression less open ended.  

```
RAW_DATE_REGEX=r'(19\d{6}$)|(20\d{6}$)'
BQ_DATE_REGEX = r'(19\d{2}-\d{2}-\d{2}$)|(20\d{2}-\d{2}-\d{2}$)'
```

This will still match some bad dates, like 2015-48-67, but it will limit the years to starting with 19 or 20.

If you wanted to make this more specific, you could use the following:

```
RAW_DATE_REGEX = r'(19\d{2}(01|02|03|04|05|06|07|08|09|10|11|12)([0-3][0-9])$)|(20\d{2}(01|02|03|04|05|06|07|08|09|10|11|12)([0-3][0-9])$)'
BQ_DATE_REGEX = r'(19\d{2}-(01|02|03|04|05|06|07|08|09|10|11|12)-([0-3][0-9])$)|(20\d{2}-(01|02|03|04|05|06|07|08|09|10|11|12)-([0-3][0-9])$)'
```
This will keep the month and years correct, but some months could have the wrong number of days.  What I'm getting at is, we shouldn't trust the data just because it comes from another project.  We need to verify it ourselves.  If we want to get really fancy, we can map out the correct months and days, too.  Then pre-compiling the regular expression becomes more important.